# SOURCES
# 1. Quick R http://www.statmethods.net/
# 2. http://manuals.bioinformatics.ucr.edu/home

# Pre-requisites
# R
# Rstudio (server, desktop)
# Bioconductor (libraries)
<<<<<<< HEAD

=======
>>>>>>> 91a7168412c0e861ef4733ff8098292f7a37e049
# git https://github.com/hgwoo/Rtutorial.git

# WEEK1
# data type
# function

<<<<<<< HEAD
# ExpressionSet = (matrix, featureData, phenoData)
# Granges (chr, start end strand, ...)

=======
# expressionSet
# Granges
>>>>>>> 91a7168412c0e861ef4733ff8098292f7a37e049

# WEEK2
# Linux command
# NGS analyis pipeline

<<<<<<< HEAD
# WEEK3
# annotation
# Differential Expression Analysis
# Clustering
=======
#WEEK3
# annotation
# Differential Expression Analysis
# clustering
>>>>>>> 91a7168412c0e861ef4733ff8098292f7a37e049
# Gene Set Analysis (with R , GSEA package)
# get GEO data

# WEEK4
# RNA-seq analysis pipeline
# visualization - ggplot
# IGV


